inference-engine/src/mkldnn_plugin/nodes/extract_image_patches.cpp

```diff
@@ -1,248 +1,425 @@
+            const int64_t ih_lpad = ih_start >= 0 ? 0 : std::ceil(- 1.f * ih_start / SH);
+            const int64_t iw_lpad = iw_start >= 0 ? 0 : std::ceil(- 1.f * iw_start / SW);
+            const int64_t ih_hpad = std::ceil((IH - 1.f * ih_start) / SH) > OH ? OH : std::ceil((IH - 1.f * ih_start) / SH);
+            const int64_t iw_hpad = std::ceil((IW - 1.f * iw_start) / SW) > OW ? OW : std::ceil((IW - 1.f * iw_start) / SW);
+
+            int64_t dst_offset = ob * ostrides[0] + kh * ostrides[1] + kw * ostrides[2] + ic * ostrides[3];
+            int64_t src_offset = ob * istrides[0] + ic * istrides[1] + ih_start * istrides[2] + iw_start + ih_lpad * SH * IW;
+
+            auto args = jit_eximpat_args();
+            args.src = src_data + src_offset * dtype_size;
+            args.dst = dst_data + dst_offset * dtype_size;
+            args.h_lo_pad = ih_lpad;
+            args.h_hi_pad = ih_hpad;
+            args.w_lo_pad = iw_lpad;
+            args.w_hi_pad = iw_hpad;
+            (*eximpat_kernel)(&args);
+        };
+        parallel_for4d(OB, KH, KW, IC, thread_body);
+    } else {
+        auto thread_body = [&](const int64_t ob, const int64_t kh, const int64_t kw, const int64_t ic) {
+            const int64_t iw_start = kw * RW - PL;
+            const int64_t ih_start = kh * RH - PT;
+            const int64_t ih_lpad = ih_start >= 0 ? 0 : std::ceil(- 1.f * ih_start / SH);
+            const int64_t iw_lpad = iw_start >= 0 ? 0 : std::ceil(- 1.f * iw_start / SW);
+
+            const int64_t ih_hpad = std::ceil((IH - 1.f * ih_start) / SH) > OH ? OH : std::ceil((IH + -1.f * ih_start) / SH);
+            const int64_t iw_hpad = std::ceil((IW - 1.f * iw_start) / SW) > OW ? OW : std::ceil((IW - 1.f * iw_start) / SW);
+
+            char *my_dst_ptr = dst_data + (ob * ostrides[0] + kh * ostrides[1] + kw * ostrides[2] + ic * ostrides[3]) * dtype_size;
+            const char *my_src_ptr = src_data + (ob * istrides[0] + ic * istrides[1] + ih_start * istrides[2] + iw_start) * dtype_size;
+
+            int64_t num_bytes_to_set = ih_lpad * OW * dtype_size;
+            memset(my_dst_ptr, 0, num_bytes_to_set);
+            my_dst_ptr += num_bytes_to_set;
+
+            const char* src_ptr_h_stop = my_src_ptr + ih_hpad * SH * IW * dtype_size;
+            for (const char *src_h_ptr = my_src_ptr + ih_lpad * SH * IW * dtype_size;
+                src_h_ptr < src_ptr_h_stop; src_h_ptr += SH * IW * dtype_size) {
+                num_bytes_to_set = iw_lpad * dtype_size;
+                memset(my_dst_ptr, 0, num_bytes_to_set);
+                my_dst_ptr += num_bytes_to_set;
+
+                const char* src_ptr_w_stop = src_h_ptr + iw_hpad * SW * dtype_size;
+                for (const char* src_w_ptr = src_h_ptr + iw_lpad * SW * dtype_size;
+                    src_w_ptr < src_ptr_w_stop; src_w_ptr += SW * dtype_size) {
+                    num_bytes_to_set = dtype_size;
+                    memcpy(my_dst_ptr, src_w_ptr, num_bytes_to_set);
```

Comment:
memcpy copies one element in fact. Could some code around be templated? One loop for example.

Reply:
You are right, it could, but probably it is better to avoid optimization of the fallback implementation (since we do not guarantee its performance) at the cost of binary size increase.